daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -213,11 +216,43 @@ case class PartialTransaction(
       }
     } yield coinst
 
+  private def lookupGlobalContract(
+      gcoid: Value.ContractId
+  ): Option[Tx.ContractInst[Value.ContractId]] =
+    gcoid match {
+      case Value.AbsoluteContractId.V1(discriminator, suffix) =>
+        globalContracts.get(discriminator).flatMap(_.get(suffix))
+      case _ =>
+        None
+    }
+
+  def lookupCachedContract(
+      coid: Value.ContractId
+  ): Option[Tx.ContractInst[Value.ContractId]] =
+    lookupLocalContract(coid).orElse(lookupGlobalContract(coid))
+
+  /** Update the globalContract if coid is an `Value.AbsoluteContractId.V1`,
+    *  idempotent otherwise.
+    */
+  def cachedContract(
+      coid: Value.ContractId,
+      contract: Tx.ContractInst[Value.ContractId]
+  ): PartialTransaction =
+    coid match {
+      case Value.AbsoluteContractId.V1(discriminator, suffix) =>
+        copy(
+          globalContracts = globalContracts.updated(
+            discriminator,
+            globalContracts.getOrElse(discriminator, Map(suffix -> contract))))
```

Comment:
This doesn't add the contract to the map if it already exists.
```suggestion
            globalContracts.getOrElse(discriminator, Map.empty) + (suffix -> contract))
```

Reply:
Good catch.